packages/devtools_app/lib/src/charts/chart_controller.dart

```diff
@@ -387,17 +291,127 @@ class ChartController extends DisposableController
     computeZoomRatio();
 
     // All tick labels need to be recompted.
-    recomputeLabels(refresh: true);
+    computeChartArea();
+    computeLabelInterval();
 
     dirty = true;
   }
 
+  void computeLabelInterval() {
+    if (zoomDuration == null && timestamps.isNotEmpty) {
+      final firstDT = DateTime.fromMillisecondsSinceEpoch(timestamps.first);
+      final lastDT = DateTime.fromMillisecondsSinceEpoch(timestamps.last);
+      final totalDuration = lastDT.difference(firstDT);
+      final totalHours = totalDuration.inHours;
+      if (totalHours == 0) {
+        final totalMinutes = totalDuration.inMinutes;
+        if (totalMinutes == 0) {
+          labelInterval = 30; // Every 30 seconds
+        } else if (totalMinutes < 10) {
+          labelInterval = 60; // Every minute
+        } else {
+          labelInterval = 120; // Every 2 minutes
+        }
+      } else if (totalHours > 0 && totalHours < 8) {
+        labelInterval = 60 * 60; // Every hour.
+      } else if (totalHours < 24) {
+        labelInterval = 60 * 60 * 4; // Every 4 hour.
+      } else {
+        labelInterval = 60 * 60 * 12; // Every 12 hours
+      }
+    } else {
+      final rangeInMinutes = zoomDuration?.inMinutes;
+      if (rangeInMinutes == null) return;
+      switch (rangeInMinutes) {
+        case 0: // Live
+          labelInterval = 20;
+          break;
+        case 1:
+          labelInterval = 15;
+          break;
+        case 5:
+          labelInterval = 60;
+          break;
+        case 10:
+          labelInterval = 120;
+          break;
+        default:
+          assert(false, 'Unexpected Duration $rangeInMinutes');
+      }
+    }
+
+    buildLabelTimestamps(refresh: true);
+  }
+
+  void buildLabelTimestamps({refresh = false}) {
+    if (isLabelsShared || timestamps.isEmpty) return;
+
+    if (refresh) {
+      labelTimestamps.clear();
+      final leftMostTimestamp = leftMostVisibleTimestampIndex;
+      final lastTimestamp = timestamps[leftMostTimestamp];
+      labelTimestamps.add(lastTimestamp);
+      var lastLabelDT = DateTime.fromMillisecondsSinceEpoch(lastTimestamp);
+      for (var index = leftMostTimestamp; index < timestampsLength; index++) {
+        final currentTimestamp = timestamps[index];
+        final currentDT = DateTime.fromMillisecondsSinceEpoch(currentTimestamp);
+        if (currentDT.difference(lastLabelDT).inSeconds >= labelInterval) {
+          labelTimestamps.add(currentTimestamp);
+          lastLabelDT = currentDT;
+        }
+      }
+
+      return;
+    }
+
+    if (labelTimestamps.isEmpty) {
+      labelTimestamps.add(timestamps.last);
+    } else {
+      // Check left label is it out of range?
+      final leftEdge = leftMostVisibleTimestampIndex;
+      if (labelTimestamps.first < timestamps[leftEdge]) {
+        // Label is outside of visible range, remove the left label.
+        labelTimestamps.removeAt(0);
```

Comment:
why is only at most one timestamp out of range?

Reply:
I guess theoretically this could happen, if the amount samples collected over the fastest label interval of 15-20 seconds was somehow not painted during that 15-20 sec.  It could happen, but I don't think that's likely, but generally 1 or 2 chart ticks a repaint happens and the live feed left-most label slowly move out of the visible chart range.

Reply:
I was just thinking that when I introduce panning the chart than this could happen.  I'll add a TODO to fix this for panning.